src/test/java/org/mockito/internal/stubbing/defaultanswers/ReturnsSmartNullsTest.java

```diff
@@ -85,4 +90,36 @@ public void should_print_the_parameters_on_SmartNullPointerException_message() t
                 .hasMessageContaining("lumpa");
         }
     }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void generic_return_type_erasure() throws Throwable {
+        class S<T> {
+            public T f() {
+                return null;
+            }
+        }
+
+        S<Boolean> s = Mockito.mock(S.class, org.mockito.Answers.RETURNS_SMART_NULLS);
```

Comment:
I thought about type tokens here, too. One problem is that they are not 100% intuitive, i.e. people might use non-resolvable types such as `TypeToken<S<? extends Number>>` or wonder about the required braces. Maybe we can offer both, it is easy enough to overload.